src/filter/bgsubtract0r/bgsubtract0r.c

```diff
@@ -0,0 +1,254 @@
+{
+  switch(param_index)
+  {
+  case 0:
+    info->name = "threshold";
+    info->type = F0R_PARAM_DOUBLE;
+    info->explanation = "Threshold for difference";
+    break;
+
+  case 1:
+    info->name = "denoise";
+    info->type = F0R_PARAM_BOOL;
+    info->explanation = "Remove noise";
+    break;
+
+  case 2:
+    info->name = "blur";
+    info->type = F0R_PARAM_DOUBLE;
+    info->explanation = "Blur alpha channel by given radius (to remove sharp edges)";
+    break;
+  }
+}
+
+void f0r_set_param_value(f0r_instance_t instance, f0r_param_t param, int param_index)
+{
+  assert(instance);
+  bgsubtract0r_instance_t* inst = (bgsubtract0r_instance_t*)instance;
+  f0r_param_color_t *c;
+
+  switch(param_index)
+  {
+  case 0:
+    inst->threshold = *((double*)param) * 255.;
+    break;
+
+  case 1:
+    inst->denoise = *((double*)param) >= 0.5;
+    break;
+
+  case 2:
+    inst->blur = (int)(*((double*)param)+0.5);
+    break;
+  }
+}
+
+void f0r_get_param_value(f0r_instance_t instance, f0r_param_t param, int param_index)
+{
+  assert(instance);
+  bgsubtract0r_instance_t* inst = (bgsubtract0r_instance_t*)instance;
+  f0r_param_color_t *c;
+
+  switch(param_index)
+  {
+  case 0:
+    *((double*)param) = (double)inst->threshold / 255.;
+    break;
+
+  case 1:
+    *((double*)param) = inst->denoise ? 1. : 0.;
+    break;
+
+  case 2:
+    *((double*)param) = inst->blur;
+    break;
+  }
+}
+
+#define MAX(x,y) ((x) > (y) ? (x) : (y));
+
+inline static uint8_t dst(uint32_t x, uint32_t y)
+{
+  uint8_t d;
+
+  d = abs((x&0xff)-(y&0xff));
+  d = MAX(d, abs(((x>>8)&0xff)-((y>>8)&0xff)));
+  d = MAX(d, abs(((x>>16)&0xff)-((y>>16)&0xff)));
+
+  return d;
+}
+
+void f0r_update(f0r_instance_t instance, double time, const uint32_t* inframe, uint32_t* outframe)
+{
+  assert(instance);
+  bgsubtract0r_instance_t* inst = (bgsubtract0r_instance_t*)instance;
+  unsigned int width = inst->width;
+  unsigned int height = inst->height;
+  unsigned int len = width * height;
+  uint8_t *mask = inst->mask;
+  int blur = inst->blur;
+  int i;
+  int j;
+  int n;
+
+  if (!inst->reference)
+  {
+    int blen = sizeof(uint32_t)*len;
+    inst->reference = malloc(blen);
+    memmove(inst->reference, inframe, blen);
+    memset(mask, 0, blen);
+  }
+  else
+  {
+    for (i=0; i<len; i++)
+      mask[i] = (dst(inst->reference[i], inframe[i]) > inst->threshold) ? 0xff : 0;
+  }
+
+  /* Clean up the mask. */
+  if (inst->denoise)
+    for (j=1; j<height-1; j++)
+      for (i=1; i<width-1; i++)
+      {
+        n = (mask[width*j+i-1]+mask[width*j+i+1]+mask[width*(j-1)+i]+mask[width*(j+1)+i]
+             + mask[width*(j-1)+i-1]+mask[width*(j-1)+i+1]+mask[width*(j+1)+i-1]+mask[width*(j+1)+i+1])/0xff;
+        if (mask[width*j+i])
+        {
+          if (n<=2) mask[width*j+i] = 0;
+        }
+        else
+        {
+          if (n>=6) mask[width*j+i] = 0xff;
+        }
+      }
+
+  for (i=0; i<len; i++)
+    outframe[i] = inframe[i] & 0xffffff | mask[i] << 24;
```

Comment:
Consider something like?
uint8_t* p;
...for...
p = &outframe[i];
p[3] = mask[i];